src/plonk/circuit.rs

```diff
@@ -1102,7 +1083,46 @@ impl<F: Field> ConstraintSystem<F> {
                 .unwrap_or(0),
         );
 
-        degree
+        std::cmp::max(degree, self.minimum_degree.unwrap_or(1))
+    }
+
+    /// Compute the number of blinding factors necessary to perfectly blind
+    /// each of the prover's witness polynomials.
+    pub fn blinding_factors(&self) -> usize {
+        // All of the prover's advice columns are evaluated at no more than
+        let factors = *self.num_advice_queries.iter().max().unwrap_or(&1);
+        // distinct points during gate checks.
+
+        // In the permutation and lookup argument the witness polynomials are
+        // evaluated at most 3 times:
+        let factors = std::cmp::max(3, factors);
+
+        // Each polynomial is evaluated at most an additional time during
+        // multiopen:
+        let factors = factors + 1;
+
+        // h(x) is derived by the other evaluations so it does not reveal
+        // anything; in fact it does not even appear in the proof.
+
+        // h(x_3) is also not revealed; the verifier only learns a single
+        // evaluation of a polynomial in x_1 which has h(x_3) and another random
+        // polynomial evaluated at x_3 as coefficients.
```

Comment:
```suggestion
        // h(x_3) is also not revealed; the verifier only learns a single
        // evaluation of a polynomial in x_1 which has h(x_3) and another
        // random polynomial evaluated at x_3 as coefficients. Specifically,
        // this is final_poly in poly::commitment::prover::create_proof.
```

Also, to check my understanding: this evaluation is derived by the verifier from `q_evals` (which we have already accounted for in the blinding calculation), but this `h(x_3)` consideration is necessary due to the additional evaluation at `x_3` for the polynomial commitment opening proof?

Reply:
Imagine a situation where the only polynomial opened at `x` is `h'(X)` -- which is what I call the polynomial produced by computing `H_0 + [x^n] H_1 + ...`. We don't want to reveal `h'(x_3)` because it leaks information that is not trivial to blind. To avoid this, we also _always_ open a polynomial `r(X)` at `x` as well, so it always ends up in the same query set as `h'(X)` does. This `r(X)` polynomial is totally random and is `random_poly_commitment` in the vanishing argument. As a result, the proof only contains the evaluation of `x_1 h'(x_3) + r(x_3)` (and perhaps other polynomials as well) which is uniformly and independently distributed from `h'(x_3)` and therefore `h'(x_3)` is not leaked.

Reply:
I instead went with adding "-- this random polynomial is "random_poly" in the vanishing argument." to the end where your suggestion was.